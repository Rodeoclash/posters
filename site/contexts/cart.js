import { createContext, useContext } from "react";

const CartContext = createContext({
  add: () => {},
  contains: () => {},
  hide: () => {},
  remove: () => {},
  show: () => {},
  showing: false,
  size: 0,
  variants: [],
});

export const useCartContext = () => useContext(CartContext);

export default CartContext;
