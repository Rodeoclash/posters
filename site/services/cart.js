export class VariantNotFound extends Error {}
export class VariantExists extends Error {}

const variantMatch = (variant1, variant2) => {
  return variant1.id === variant2.id;
};

const variantNoMatch = (variant1, variant2) => {
  return !variantMatch(variant1, variant2);
};

export const show = (cart) => {
  return {
    ...cart,
    showing: true,
  };
};

export const hide = (cart) => {
  return {
    ...cart,
    showing: false,
  };
};

export const contains = (cart, variant) => {
  return cart.variants.some(variantMatch.bind(this, variant));
};

export const add = (cart, variant) => {
  if (contains(cart, variant) === true) {
    throw new VariantExists();
  }

  return {
    ...show(cart),
    variants: cart.variants.concat([variant]),
  };
};

export const remove = (cart, variant) => {
  if (contains(cart, variant) === false) {
    throw new VariantNotFound();
  }

  return {
    ...show(cart),
    variants: cart.variants.filter(variantNoMatch.bind(this, variant)),
  };
};
