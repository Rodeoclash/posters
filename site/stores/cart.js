import { createContext } from "react";
import { useObservable } from "mobx-react-lite";

const CartContext = createContext();

function Cart({ children }) {
  const store = useObservable({ counter: 0 });
  return <CartContext.Provider value={store}>{children}</CartContext.Provider>;
}

export { Cart, CartContext };
