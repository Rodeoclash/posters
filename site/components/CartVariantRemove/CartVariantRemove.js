import { useCartContext } from "../../contexts/cart";

import styles from "./CartVariantRemove.module.css";

const CartVariantRemove = ({ variant }) => {
  const { remove, contains } = useCartContext();

  if (contains(variant) === false) {
    return null;
  }

  return <button onClick={remove.bind(this, variant)}>Remove</button>;
};

export default CartVariantRemove;
