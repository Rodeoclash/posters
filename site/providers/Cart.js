import { useState } from "react";
import CartContext from "../contexts/cart.js";
import * as cart from "../services/cart";

const CartProvider = ({ children }) => {
  const [cartState, setCartState] = useState({
    variants: [],
    showing: false,
  });

  const add = (variant) => {
    setCartState(cart.add(cartState, variant));
  };

  const contains = (variant) => {
    return cart.contains(cartState, variant);
  };

  const hide = () => {
    setCartState(cart.hide(cartState));
  };

  const remove = (variant) => {
    setCartState(cart.remove(cartState, variant));
  };

  const show = () => {
    setCartState(cart.show(cartState));
  };

  const value = {
    add,
    contains,
    hide,
    remove,
    show,
    showing: cartState.showing,
    size: cartState.variants.length,
    variants: cartState.variants,
  };

  return <CartContext.Provider value={value}>{children}</CartContext.Provider>;
};

export default CartProvider;
