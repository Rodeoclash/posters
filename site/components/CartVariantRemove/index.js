import CartVariantRemove from "./CartVariantRemove";
export default CartVariantRemove;
