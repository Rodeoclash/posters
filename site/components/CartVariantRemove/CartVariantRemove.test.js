import React from "react";
import renderer from "react-test-renderer";
import { fireEvent, render } from "@testing-library/react";

import * as CartContext from "../../contexts/cart";
import CartVariantRemove from "./CartVariantRemove";

describe("onClick", () => {
  it("calls function to remove variant from cart", () => {
    const variant = {};
    const remove = jest.fn();

    const contextValues = {
      contains() {
        return true;
      },
      remove,
    };

    jest
      .spyOn(CartContext, "useCartContext")
      .mockImplementation(() => contextValues);

    const { getByText } = render(<CartVariantRemove variant={variant} />);

    fireEvent.click(getByText("Remove"));

    expect(remove).toHaveBeenCalledWith(variant, expect.anything());
  });
});

describe("snapshots", () => {
  it("with variant in the cart", () => {
    const variant = {};
    const remove = jest.fn();

    const contextValues = {
      contains() {
        return true;
      },
      remove,
    };

    jest
      .spyOn(CartContext, "useCartContext")
      .mockImplementation(() => contextValues);

    const tree = renderer
      .create(<CartVariantRemove variant={variant} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("with variant not in the cart", () => {
    const variant = {};
    const remove = jest.fn();

    const contextValues = {
      contains() {
        return false;
      },
      remove,
    };

    jest
      .spyOn(CartContext, "useCartContext")
      .mockImplementation(() => contextValues);

    const tree = renderer
      .create(<CartVariantRemove variant={variant} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
