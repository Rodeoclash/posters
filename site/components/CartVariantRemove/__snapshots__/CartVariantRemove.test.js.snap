// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots with variant in the cart 1`] = `
<button
  onClick={[Function]}
>
  Remove
</button>
`;

exports[`snapshots with variant not in the cart 1`] = `null`;
